qiskit/transpiler/passes/optimization/commutative_cancellation.py

```diff
@@ -61,7 +63,18 @@ def __init__(self, basis_gates=None, target=None):
             self.basis = set(target.operation_names)
 
         self._var_z_map = {"rz": RZGate, "p": PhaseGate, "u1": U1Gate}
-        self.requires.append(CommutationAnalysis())
+
+        self._z_rotations = {"p", "z", "u1", "rz", "t", "s"}
+        self._x_rotations = {"x", "rx"}
+        self._gates = {"cx", "cy", "cz", "h", "y"}  # Now the gates supported are hard-coded
```

Comment:
That's a great PR!
Does this pass only checks self commuting gates? or also pairs of gates, for example:
- Any 1q gate in _z_rotations commutes with the control of CX
- Any 1q gate in _x_rotations commutes with the target of CX
- Any  1q gate in _z_rotations commutes with the control and target of CZ

Reply:
Regarding CX, the pass does the following:
```
          ┌───┐   ┌───┐   ┌─────────┐
q_0: ──■──┤ S ├───┤ T ├───┤ Rz(2.1) ├──■──
     ┌─┴─┐├───┤┌──┴───┴──┐└─────────┘┌─┴─┐
q_1: ┤ X ├┤ X ├┤ Rx(1.2) ├───────────┤ X ├
     └───┘└───┘└─────────┘           └───┘
     ┌────────────┐
q_0: ┤ Rz(4.4562) ├
     ├────────────┤
q_1: ┤ Rx(4.3416) ├
     └────────────┘
```
and then same for CZ. However, the pass could use some updating as it cannot handle e.g. `Sdg` or `Tdg` gates, meaning that the following circuit is unaffected by the pass
```
          ┌─────┐
q_0: ──■──┤ Sdg ├──■──
     ┌─┴─┐└─────┘┌─┴─┐
q_1: ┤ X ├───────┤ X ├
     └───┘       └───┘
```
In practice that is caught by re-synthesizing 2q unitaries but we could do it cheaper with the commutative cancellation. Let's do that in a follow up and also add support for SX gates.

Reply:
why not to add `sdg` and `tdg` to `self._z_rotations` ?

Reply:
That's exactly what to do, but I'd rather do it in a follow up and not change the functionality so close to the 1.2 release. I can open the follow-up the next days and we can discuss separately 🙂